src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/auth/OAuth1aRequestFilter.java

```diff
@@ -0,0 +1,333 @@
+import net.oauth.OAuthMessage;
+import net.oauth.OAuthProblemException;
+import net.oauth.OAuthValidator;
+import net.oauth.server.OAuthServlet;
+
+import javax.annotation.CheckForNull;
+import javax.inject.Inject;
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Clock;
+import java.util.Map;
+import java.util.Optional;
+import java.util.logging.Logger;
+
+import static com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.OAuthRequestUtils.isOAuthAccessAttempt;
+import static com.atlassian.bitbucket.jenkins.internal.applink.oauth.util.OAuthProblemUtils.logOAuthProblem;
+import static com.atlassian.bitbucket.jenkins.internal.applink.oauth.util.OAuthProblemUtils.logOAuthRequest;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.String.format;
+import static java.util.logging.Level.*;
+import static javax.servlet.RequestDispatcher.FORWARD_REQUEST_URI;
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+import static net.oauth.OAuth.Problems.*;
+import static net.oauth.OAuthMessage.AUTH_SCHEME;
+import static org.apache.commons.lang3.StringUtils.startsWithIgnoreCase;
+import static org.apache.http.HttpHeaders.AUTHORIZATION;
+
+public class OAuth1aRequestFilter implements Filter {
+
+    private static final Logger log = Logger.getLogger(OAuth1aRequestFilter.class.getName());
+
+    private final ConsumerStore consumerStore;
+    private final ServiceProviderTokenStore tokenStore;
+    private final OAuthValidator validator;
+    private final Clock clock;
+    private final TrustedUnderlyingSystemAuthorizerFilter authorizerFilter;
+
+    @Inject
+    public OAuth1aRequestFilter(ConsumerStore consumerStore,
+                                ServiceProviderTokenStore tokenStore,
+                                OAuthValidator validator,
+                                Clock clock,
+                                TrustedUnderlyingSystemAuthorizerFilter authorizerFilter) {
+        this.consumerStore = consumerStore;
+        this.tokenStore = tokenStore;
+        this.validator = validator;
+        this.clock = clock;
+        this.authorizerFilter = authorizerFilter;
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response,
+                         FilterChain chain) throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+
+        if (isOauthRequest(req)) {
+            OAuthMessage message = OAuthServlet.getMessage(req, getLogicalUri(req));
+            String tokenStr = getTokenFromRequest(req, resp, message);
+            if (tokenStr == null) {
+                return;
```

Comment:
nit: are these explicit `return`s necessary?

Reply:
I would prefer those since it short circuits while someone is reading the code.